feat: added lifecycle color-coded text

akshatnema opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

This PR adds a component named `LifeCycle` which provides color-coded text to Lifecycle terms. It takes 3 keywords and outputs the text in a specific color, according to the keyword added. Till now, 3 keywords have been added. They are: `Deprecated`, `Required` and `Optional`. 

**Related issue(s)**
Resolves #592 
<!-- If you refer to a particular issue, provide its number, othewise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->

akshatnema committed:
added lifecycle color text

akshatnema committed:
removed unnecesarry color text

akshatnema commented:
@imabp WIll you please go through the preview link and check whether it is placed at the right positions?

imabp commented on tailwind.config.js:
```diff
@@ -39,6 +39,9 @@ module.exports = {
       },
+      'deprecated': '#E81717',
+      'required': '#3B6B8E',
+      'optional': '#CECCCC'
```
This should be under `extended:{colors:{}}`

akshatnema committed:
make required to yellow

akshatnema committed:
updated approach

akshatnema committed:
Merge branch 'master' into akshat-branch-1

imabp commented:
@akshatnema the background colors are still not visible

akshatnema committed:
Merge branch 'master' into akshat-branch-1

akshatnema committed:
setting purge to true

akshatnema committed:
Merge branch 'akshat-branch-1' of https://github.com/akshatnema/website into akshat-branch-1

akshatnema committed:
changing purge back to false

akshatnema committed:
Merge branch 'master' into akshat-branch-1

akshatnema committed:
added in postcss config

fmvilas commented on pages/docs/specifications/v2.0.0.md:
```diff
@@ -96,11 +96,11 @@ It combines resource listing and API declaration together into one document.
 Field Name | Type | Description
 ---|:---:|---
-<a name="A2SAsyncAPI"></a>asyncapi | [AsyncAPI Version String](#A2SVersionString) | **Required.** Specifies the AsyncAPI Specification version being used. It can be used by tooling Specifications and clients to interpret the version. The structure shall be `major`.`minor`.`patch`, where `patch` versions _must_ be compatible with the existing `major`.`minor` tooling. Typically patch versions will  ...
+<a name="A2SAsyncAPI"></a>asyncapi | [AsyncAPI Version String](#A2SVersionString) | <LifeCycle text="required" /> Specifies the AsyncAPI Specification version being used. It can be used by tooling Specifications and clients to interpret the version. The structure shall be `major`.`minor`.`patch`, where `patch` versions _must_ be compatible with the existing `major`.`minor` tooling. Typically patc ...
```
You can't use MDX here. The spec files are [automatically copied from the spec repo](https://github.com/asyncapi/spec/blob/master/.github/workflows/new-spec-release.yml).

We're using MDX in the website but not in other parts so this file must remain pure Markdown.

imabp replied:
@fmvilas what you think about running a script post loading of a page,  which checks the dom nodes with the text content matching to the life cycle parameters....  And adds span tags with classes around it.

fmvilas replied:
If you manage to make it, I'm all ears 😄 The question is: is it worth the effort? :)

imabp replied:
I dont think so, and it will take good efforts. 😅
- Its just a small class change.
- Not having much value addition.

akshatnema replied:
>You can't use MDX here. The spec files are [automatically copied from the spec repo](https://github.com/asyncapi/spec/blob/master/.github/workflows/new-spec-release.yml).

Yupp, I understand that it is not an MDX but we are still using many of the components in `docs` and `blogs` to visualize the components. And there can't be text color coding in MD files manually by adding `div` tag. If the custom CSS classes don't work in this md, we surely have to go for another implementation.

>@fmvilas what you think about running a script post loading of a page, which checks the dom nodes with the text content matching to the life cycle parameters.... And adds span tags with classes around it.

This is BTW Nice idea to implement but I don't think it's reliable. It is because `REQUIRED` and `OPTIONAL` are not only life cycle parameters, they are also normal language words used to define certain parameters in normal English. If we made a script, it will take all the words naming `DEPRECATED`, `REQUIRED` and `OPTIONAL` from the spec file and will make it highlighted. There should be any way to differentiate the lifecycle terms from normal English using words.

fmvilas reviewed (changes requested):
Left a comment. I think it may not be possible to implement what you're trying to do here :/

akshatnema committed:
Merge branch 'master' into akshat-branch-1

akshatnema committed:
updated colors

akshatnema commented:
@imabp Now, the background colors are visible. There is a slight change in the color. Hope it doesn't matter to you. Please review it.

akshatnema commented:
I also request @fmvilas @alequetzalli to kindly review this PR and do let me know if I have missed any lifecycle terms or have added somewhere inappropriately. Thanks.

fmvilas commented on components/LifeCycle.js:
```diff
@@ -0,0 +1,11 @@
+export default function LifeCycle({text, classname=''}) {
+  let Color, bgColor;
```
Please be consistent with style:

```suggestion
  let color, bgColor;
```

fmvilas commented on components/LifeCycle.js:
```diff
@@ -0,0 +1,11 @@
+export default function LifeCycle({text, classname=''}) {
+  let Color, bgColor;
+  let text_type=text.toUpperCase();
```
Same as above, either use camelCase or snake_case but not both. We're using camelCase usually.

```suggestion
  let textType=text.toUpperCase();
```

fmvilas commented on components/LifeCycle.js:
```diff
@@ -0,0 +1,11 @@
+  let text_type=text.toUpperCase();
+  if(text_type==="DEPRECATED") {Color="white", bgColor="red-500"}
+  else if(text_type==="REQUIRED") {Color="black", bgColor="yellow-300"}
+  else if(text_type==="OPTIONAL") {Color="black", bgColor="gray-300"}
```
Please use single quotes. I have mixed feelings about using the comma operator this way too but leave it up to you to leave it or change it.

fmvilas commented on pages/blog/release-notes-2.1.0.md:
```diff
@@ -18,7 +18,7 @@ The last AsyncAPI release (2.0.0) took place on the 11th of September, 2019. In
 ## Message examples object extended with additional fields
 
-Thanks to work done by [Laurent Broudoux](https://github.com/lbroudoux), you can now clearly describe message examples. New properties **name** and **summary** are optional. These properties help not only to properly describe the example in documentation but make it easier to work with mocking and testing tools (like [microcks](https://microcks.io/)), so you can better identify what example to us ...
+Thanks to work done by [Laurent Broudoux](https://github.com/lbroudoux), you can now clearly describe message examples. New properties **name** and **summary** are <LifeCycle text="optional" />. These properties help not only to properly describe the example in documentation but make it easier to work with mocking and testing tools (like [microcks](https://microcks.io/)), so you can better identi ...
```
This is not accessible. People using screen readers will not hear the "optional" word. We should either add `aria-label` attribute or simply wrap the text:

```suggestion
Thanks to work done by [Laurent Broudoux](https://github.com/lbroudoux), you can now clearly describe message examples. New properties **name** and **summary** are <LifeCycle>optional</LyfeCycle>. These properties help not only to properly describe the example in documentation but make it easier to work with mocking and testing tools (like [microcks](https://microcks.io/)), so you can better identify what example to use for mocking and what it does.
```

fmvilas commented on components/LifeCycle.js:
```diff
@@ -0,0 +1,11 @@
+export default function LifeCycle({text, classname=''}) {
```
The name LifeCycle says nothing IMHO. Also `required` and `optional` have nothing to do with the life cycle. I think this is the perfect example of a component that is trying to do too much. My recommendation is that you split it into 3 components: `Deprecated`, `Required`, and `Optional`. This way, people can use like this:

```mdx
something something is <Required>required</Required> something something
```

That will only apply certain styles but will not alter the text inside. For instance, if you want everything to be upper case, you do it using the `text-uppercase` CSS class instead of transforming the text. You also don't have to check for the content to match "deprecated" or anything, just wrap it inside a span element with certain CSS classes. It's up to the person using this component to use it wisely.

fmvilas reviewed (changes requested):
Left a bunch of comments. This PR is one of those examples that initially look innocent but end up having lots of side effects 😄 

BTW, you forgot to remove the changes you did in the specification documents. Please, revert them.

akshatnema committed:
Spilt of Component into 3

akshatnema committed:
created 3 components

akshatnema commented on pages/blog/release-notes-2.2.0.md:
```diff
@@ -60,4 +60,4 @@ Are you wondering how we managed to release 2.2.0 just three months after 2.1.0?
 
 Does the above meme give you mixed feelings? Are you afraid of possible changes, or actually happy to see it coming? Don't overthink it! Join our [Slack](https://www.asyncapi.com/slack-invite) and talk to us, or check out the [3.0.0 milestone](https://github.com/asyncapi/spec/milestone/18).
 
-> Photo by <a href="https://unsplash.com/@jeremythomasphoto?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Jeremy Thomas</a> on <a href="https://unsplash.com/s/photos/autumn?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>
```
I really don't know what's the change here. Maybe there is a single character difference in the file but I'm not getting the exact diff here. Hope that you won't consider it much :slightly_smiling_face:. Same goes for rest specification file and release notes file.

akshatnema commented:
> My recommendation is that you split it into 3 components: Deprecated, Required, and Optional

@fmvilas Yepp, I have changed it to this implementation now, since it makes it much more flexible to add the relevant component with required formatting anywhere in the file.

> BTW, you forgot to remove the changes you did in the specification documents. Please, revert them.

Hmm, I did it before because I thought it is part of my job. But no problems :v:, I have reverted them. If you want to do it from your side or using spec generation scripts, it will be great. 

The rest of the changes required have been completely removed from the PR. So, they are no longer needed to be corrected :smile:. Thanks for your review and feedback. If certain more changes are to be implemented, feel free to tell me. I'm there to go with them.

fmvilas reviewed (changes requested):
🤔 You split the component in 3 but I don't see you using it anywhere. Why are you creating them then?

akshatnema commented:
> 🤔 You split the component in 3 but I don't see you using it anywhere. Why are you creating them then?

@fmvilas That's what I had done before and you asked me to revert the changes in specification files, so I reverted them. If you think I need to add it again, I will do it again. It's better if you can guide me where it is needed.

fmvilas commented:
I think we're not understanding each other. See my comment above: https://github.com/asyncapi/website/pull/597#discussion_r809400240. You can't use MDX because spec files are copied automatically from [spec repo](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md). This file can't have MDX syntax because it's used in other places where MDX is not supported + it's a normative document.

I thought you wanted to use these components in other places but if it's not the case I don't think this PR is needed then.

akshatnema commented:
@fmvilas I understand that we can't use MDX directly in specification files as they are generated from scripts. But if certain special encoded text is needed to show lifecycle terms in specification files, what can be another approach/implementation to do this?

If this is not needed, we can surely close this PR and issue #592 to avoid confusion for other contributors and community members. 

So sorry, I interpreted your message incorrectly 😅.

fmvilas commented:
> what can be another approach/implementation to do this?

You have almost no options AFAIK. @imabp pointed out a solution where you scan all the text and wrap all the "optional" words in an Optional component but, as you said, it's not trivial.

> If this is not needed, we can surely close this PR and issue https://github.com/asyncapi/website/issues/592 to avoid confusion for other contributors and community members.

IMHO, I don't think this is needed. It's nice to have but definitely not something we need. So yeah, I'd close the PR but happy to listen to other opinions. cc @alequetzalli @derberg 

> So sorry, I interpreted your message incorrectly 😅.

No need to apologize mate. I actually feel bad you spent some time working on something that probably will end up not being used.

quetzalliwrites commented:
Interesting! This def ended up being more complex than it seemed. 😄 

I personally don't think we MUST have it, but I do think it could be useful should it be implemented. 

Regrettably, I don't know of another technical solution to implement it other than what has already been discussed in this PR so far.

akshatnema commented:
No Problem at all. We know that it is a good change but the only thing is that the implementation requires a lot of effort (which is not worth it, comparing the change we are producing). I request @fmvilas to close that issue as well if you think it can't be implemented or it is not required. 

Meanwhile, anyone who wishes to provide some closing notes to this PR, feel free to do that as I am getting an option of `Close with comment` :smile:.

fmvilas commented:
No comments from my side. I think it's clear in the discussions above :) Sorry, @akshatnema! And thanks a lot for pushing it 🙏
